python/servo/bootstrap.py

```diff
@@ -13,26 +13,55 @@
 
 import servo.packages as packages
 from servo.util import extract, download_file, host_triple
+from mach_bootstrap import _get_virtualenv_script_dir
 
 
-def salt(context, force=False):
+def run_as_root(command):
+    if os.geteuid() != 0:
+        try:
+            # Check if sudo'ing possible
+            subprocess.check_call(['sudo', '-v'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            command.insert(0, 'sudo')
+        except:
+            command = ['su', 'root', '-c', ' '.join(command)]
+
+    return subprocess.call(command)
+
+
+# This install needed dependencies for Salt
+def install_salt_dependencies(context, force):
+    print("Installing missing Salt dependencies...")
+    if context.distro == 'Ubuntu':
+        run_as_root(['apt-get', 'install', '-y', 'build-essential', 'libssl-dev', 'libffi-dev', 'python-dev'])
+    elif context.distro in ('CentOS', 'CentOS Linux', 'Fedora'):
+        run_as_root(['dnf', 'install', '-y', 'gcc', 'libffi-devel', 'python-devel', 'openssl-devel'])
+
+    return salt(context, force, retry=True)
+
+
+def salt(context, force=False, retry=False):
     # Ensure Salt is installed in the virtualenv
     # It's not instaled globally because it's a large, non-required dependency,
     # and the installation fails on Windows
-    print("Checking Salt installation...", end='')
-    reqs_path = os.path.join(context.topdir, 'python', 'requirements-salt.txt')
-    process = subprocess.Popen(
-        ["pip", "install", "-q", "-I", "-r", reqs_path],
-        stdout=subprocess.PIPE,
-        stderr=subprocess.PIPE
-    )
-    process.wait()
-    if process.returncode:
-        out, err = process.communicate()
-        print('failed to install Salt via pip:')
-        print('Output: {}\nError: {}'.format(out, err))
-        return 1
-    print("done")
+    virtualenv_script_dir = os.path.join(context.topdir, 'python', '_virtualenv', _get_virtualenv_script_dir())
```

Comment:
It would be nice to add an `absolute=True` (optional, kwarg?) arg to `_get_virtualenv_script_dir` so thatthe path prefixes don't need to be appended here and in `mach_bootstrap.py`.